host/ecallids.c

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "ecallids.h"
+#include <openenclave/internal/raise.h>
+#include "hostthread.h"
+
+/* The maximum numer of different ecalls that an application can make is
+ * bounded. */
+static const char* _ecall_table[OE_MAX_ECALLS];
+static uint32_t _ecall_table_size;
+
+/* Mutex for assigning/looking up global ids in a thread-safe manner. */
+static oe_mutex _lock;
+
+/* Get the global ecall id from the _ecall_table. Locking must be done by
+ * caller. */
+static oe_result_t _get_global_id(const char* name, uint64_t* global_id)
+{
+    oe_result_t result = OE_NOT_FOUND;
+    if (!name || !global_id)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Search for id assigned for given name. */
+    for (uint64_t i = 0; i < _ecall_table_size; i++)
+    {
+        if (strcmp(_ecall_table[i], name) == 0)
+        {
+            *global_id = i;
+            break;
+        }
+    }
+
+    /* If the name is not found, adding it to the table. */
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+    {
+        if (_ecall_table_size >= OE_MAX_ECALLS)
+        {
+            *global_id = OE_MAX_ECALLS;
```

Comment:
seems safer to pass back OE_GLOBAL_ECALL_ID_NULL rather than OE_MAX_ECALLS

Reply:
ID_NULL indicates that the global_id has not been initialalized yet while ID_MAX indicates the global_id has been initialalized but failed. In this case, the corresponding ecall will always fail. We just reuse the ID_MAX as error code. Alternatively, we can introduce something like ID_ERROR.

Reply:
I don't see why we need two different values.
I also don't think having any hard coded max.

Reply:
In the latest version, the function returns `OE_NOT_FOUND` instead of `OE_OK`. The is more accurate since no implementation of the given ecall was found for the given enclave.